internal/graph/check.go

```diff
@@ -103,6 +124,8 @@ func NewLocalChecker(
 	concurrencyLimit uint32,
 ) *LocalChecker {
 	checker := &LocalChecker{ds: ds, concurrencyLimit: concurrencyLimit}
+	checker.resolver = checker
```

Comment:
can you add a few lines explaining this?  I have a hard time wrapping my head around the notion that the checker's resolver is itself.

Reply:
+1. This exact comment came up in the original Check rewrite: https://github.com/openfga/openfga/pull/476#discussion_r1089530447

> can you add a few lines explaining this?

I personally don't think this is enough.. this may be a solution that works and is clever but it's not readable.

Reply:
> +1. This exact comment came up in the original Check rewrite: https://github.com/openfga/openfga/pull/476#discussion_r1089530447

Let's not forget context. The original intent behind the comment was motivated by the fact that the term "dispatching" was a conflicting topic. At the time we only need a local concurrent CheckResolver, and so the extra wrapped `resolver` was a bit much to swallow at the time. The original code was already written to take into account what this PR introduces. Fact of the matter is, dispatch composition has many uses, and this is one of them.

I don't understand what's so terribly wrong with it? Could you elaborate and be more specific? It may be a bit inception(y), but there isn't anything inherently wrong with it. 

The primary driving factor behind this is that we're delegating to an implementation of an interface, and the interface you happen to be delegating to is yourself if you are only constructing a `LocalChecker`. But since the inner resolver can be dispatched through any implementation of a CheckResolver, then it allows for clever compositional routing through an alternative implementation of a CheckResolver, which for the scope of this PR is a concurrency limited resolver but in the future it could be a remote dispatcher etc..

Reply:
This is a different way to look at it in simpler terms that may help to understand the _why_ behind the pattern here..

```go
type SomeInterface interface {
  SomeFunc()
}

type SomeStruct struct {
}

func (s *SomeStruct) SomeFunc() {...}

func NewSomeStruct() *SomeStruct {
    return &SomeStruct{}
}

s := NewSomeStruct()
s.SomeFunc()
```
is functionally equivalent to

```go
type SomeInterface interface {
  SomeFunc()
}

type SomeStruct struct {
    delegate SomeInterface
}

func (s *SomeStruct) SomeFunc() {s.delegate.SomeFunc()}

func NewSomeStruct() *SomeStruct {
    s := &SomeStruct{}
    s.delegate = s
    return s
}

s := NewSomeStruct()
s.SomeFunc()
```
The difference is that the first cannot benefit from functional composition chaining whereas the second can. If you have more than one implementation of `SomeInterface`, then you can chain/compose the delegate with those, and that allows for decorating and chaining the interface implementations.

Reply:
@miparnisari mind taking a stab at what you would do differently?

From a diff standpoint, I think the top goal for me is just to avoid too many changes to the LocalChecker. But if we can do it differently that achieves the same outcome then let's do it.

Reply:
I am not arguing whether it is correct or not correct (likely correct).  However, the fact that it feel `inception(y)` to both @miparnisari and myself suggests we can explain in the code as comments what it does (or why we have resolver = checker) in that it is a dispatching pattern.  This will help others reading the code wrapping their head around it.

Reply:
@adriantam I have no problem adding a comment. @miparnisari seemed to suggest we change it, and that was the motivation of my responses above.

I'll let ya'll take a stab at it and fix it up as you see fit. This was just to serve as an example.

Reply:
I'll post a link here to another version... when/if I actually do get a working version 😆